libff/algebra/curves/bls12_381/bls12_381_pairing.cpp

```diff
@@ -0,0 +1,473 @@
+    c.ell_VW = bls12_381_twist * D;                                // ell_VW = D (later: * yP    )
+}
+
+bls12_381_ate_G1_precomp bls12_381_ate_precompute_G1(const bls12_381_G1& P)
+{
+    enter_block("Call to bls12_381_ate_precompute_G1");
+
+    bls12_381_G1 Pcopy = P;
+    Pcopy.to_affine_coordinates();
+
+    bls12_381_ate_G1_precomp result;
+    result.PX = Pcopy.X;
+    result.PY = Pcopy.Y;
+
+    leave_block("Call to bls12_381_ate_precompute_G1");
+    return result;
+}
+
+bls12_381_ate_G2_precomp bls12_381_ate_precompute_G2(const bls12_381_G2& Q)
+{
+    enter_block("Call to bls12_381_ate_precompute_G2");
+
+    bls12_381_G2 Qcopy(Q);
+    Qcopy.to_affine_coordinates();
+
+    bls12_381_Fq two_inv = (bls12_381_Fq("2").inverse()); // could add to global params if needed
+
+    bls12_381_ate_G2_precomp result;
+    result.QX = Qcopy.X;
+    result.QY = Qcopy.Y;
+
+    bls12_381_G2 R;
+    R.X = Qcopy.X;
+    R.Y = Qcopy.Y;
+    R.Z = bls12_381_Fq2::one();
+
+    const bigint<bls12_381_Fq::num_limbs> &loop_count = bls12_381_ate_loop_count;
+    bool found_one = false;
+    bls12_381_ate_ell_coeffs c;
+
+    for (long i = loop_count.max_bits(); i >= 0; --i)
+    {
+        const bool bit = loop_count.test_bit(i);
+        if (!found_one)
+        {
+            /* this skips the MSB itself */
+            found_one |= bit;
+            continue;
+        }
+
+        doubling_step_for_miller_loop(two_inv, R, c);
+        result.coeffs.push_back(c);
+
+        if (bit)
+        {
+            mixed_addition_step_for_miller_loop(Qcopy, R, c);
+            result.coeffs.push_back(c);
+        }
+    }
+
+    leave_block("Call to bls12_381_ate_precompute_G2");
+    return result;
+}
+
+bls12_381_Fq12 bls12_381_ate_miller_loop(const bls12_381_ate_G1_precomp &prec_P,
+                                     const bls12_381_ate_G2_precomp &prec_Q)
+{
+    enter_block("Call to bls12_381_ate_miller_loop");
+
+    bls12_381_Fq12 f = bls12_381_Fq12::one();
+
+    bool found_one = false;
+    size_t idx = 0;
+
+    const bigint<bls12_381_Fq::num_limbs> &loop_count = bls12_381_ate_loop_count;
+    bls12_381_ate_ell_coeffs c;
+
+    for (long i = loop_count.max_bits(); i >= 0; --i)
+    {
+        const bool bit = loop_count.test_bit(i);
+        if (!found_one)
+        {
+            /* this skips the MSB itself */
+            found_one |= bit;
+            continue;
+        }
+
+        /* code below gets executed for all bits (EXCEPT the MSB itself) of
+           bls12_381_param_p (skipping leading zeros) in MSB to LSB
+           order */
+
+        c = prec_Q.coeffs[idx++];
+        f = f.squared();
+        f = f.mul_by_045(c.ell_0, prec_P.PY * c.ell_VW, prec_P.PX * c.ell_VV);
+
+        if (bit)
+        {
+            c = prec_Q.coeffs[idx++];
+            f = f.mul_by_045(c.ell_0, prec_P.PY * c.ell_VW, prec_P.PX * c.ell_VV);
+        }
+
+    }
+
+    if (bls12_381_ate_is_loop_count_neg)
+    {
+    	f = f.inverse();
+    }
```

Comment:
This block is always executed for BLS12-381 (but won't be for BLS12-377). What about removing the `if` here - and re-introducing back when the code is refactored to support "BLS12" as a family?

Reply:
I suggest to keep it consistent with `alt_bn128` where the seed is also positive, and still there is the `if` block . It would be "easier" then to refactor into families.

Reply:
>  I suggest to keep it consistent with alt_bn128 where the seed is also positive, and still there is the if block

Oh that's right. Makes sense to say consistent then.

> It would be "easier" then to refactor into families.

Yes, this conditional will truly be meaningful when curve families are implemented :)